distributed/shuffle/shuffle-design.md

```diff
@@ -0,0 +1,253 @@
+
+Additionally, cleanly stopping a running shuffle takes more than just releasing the shuffle tasks from memory: since there's out-of-band processing going on, the `ShuffleExtension` has to be informed in some way that it needs to stop doing whatever it's doing in the background, and clear out its buffers. Also, executing tasks may be blocking on the `ShuffleExtension` doing something; without a way to tell the extension to shut down, those tasks might block forever, deadlocking the cluster.
+
+Therefore, we propose adding a `RerunGroup` (`ImpureGroup`? `CoExecutionGroup`? `RestartGroup`? `OutOfBandGroup`? name TBD) structure to the scheduler which intertwines the fates of all tasks within it: if any one task is to be rescheduled (due to its worker leaving), all tasks are restarted; if any one is to be prematurely released (due to cancellation), all are released. (xref [dask/distributed#5403](https://github.com/dask/distributed/issues/5403))
+
+Membership in a `RerunGroup` is implemented via task annotations, where each task gives the name of the `RerunGroup` it belongs to. A task can belong to at most one `RerunGroup`. TBD if we will enforce any structural restrictions on `RerunGroup`s to prevent odd/invalid states from emerging—we probably should, such as not allowing disjoint parts of the graph to be in the same `RerunGroup`, etc. (Note that fixes to annotations will be required for this to work, particularly clear rules for when and how annotated layers can be fused xref [dask/dask#7036](https://github.com/dask/dask/issues/7036).)
+
+Additionally, the scheduler informs workers whenever a `RerunGroup` is restarted or cancelled. Workers will have a way to pass this information on to any interested out-of-band operations. This could be something like:
+- Workers have a named `threading.Event` for each `RerunGroup` that any of their current tasks belong to. When the scheduler tells workers about a restart/cancellation, they `set()` the corresponding event so that some background thread can respond accordingly.
+- A `register_cancellation_handler(rerun_group: str, async_callback: Callable)` method on workers that registers an async function to be run when that group is cancelled/restarted. A potential upside (and potential deadlock) is that the scheduler's `cancel_rerun_group` RPC to workers could block on this callback completing, meaning the scheduler wouldn't treat the `RerunGroup` as successfully cancelled until every callback on every worker succeeded. That could give us some nice synchronization guarantees (which we may or many not actually need) ensuring a shuffle doesn't try to restart while it's also trying to shut down.
+
+### Peer discovery and initialization
```

Comment:
> Say you have 100 workers and 2 shuffles. The two shuffle_inits run on workers 0 and 1. Then worker 99 leaves. shuffle_init would not be rescheduled, because no keys in the RerunGroup were lost.

That assumes that the `RerunGroup` triggers on released key. We could just as well trigger it on `remove_worker` and there is no problem.

> We should probably suck it up and create a scheduler plugin for shuffling.

Indeed this is something I would've started with. I don't consider it an additional layer of complexity since it simplifies the graph. However, we still need to do the same thing. Whether we put this into the shuffle-setup, the scheudler plugin, the barrier, etc. we still need to 
* generate a unique shuffle id
* release all keys if something goes wrong (e.g. error, worker lost)
* set output restrictions
* ???
In my mind the only new (is it actually new??) requirement is to (trigger a `RerunGroup` | restart a shuffle) if a worker is lost. The plugin is a different way to do it and it might have synergies with other requirements (e.g. if we can avoid another complex optimization step and deal with culling more sensibly, I'm all in favour of the plugin)

I wouldn't reject the idea of a `RerunGroup` entirely based on this. However, I see the appeal of a SchedulerPlugin

Re: barrier task

> [Excalidraw] no-op dummy task to create a dependency between `unpack` and `transfer` and enforce task ordering. In theory, we could just have an all-to-all graph like this [...] but that would be much more burden on the scheudler to handle.

Just a technical comment since I was thrown off by this as well for a second. This is not a huge burden for the scheduler. The scheduler doesn't care (a lot) about the edges we have in our graph.
Edges for the scheduler merely mean
* dask.order is computationally more expensive (that might be sizeable)
* A task transition causes some set.remove/discard but I consider this cheap even for large sets
Otherwise edges mean that the `compute-task` signal for `unpack` includes a list of all workers in the `who_has` key but that's also relatively cheap.

I assume the biggest factor will be that every worker will again need to connect to every worker to fetch the result of every `transfer`. While the result is likely None, the scheduler will still open a comm to every worker. upon receiving this key, it will let the scheduler know about it with `add-keys`. We also need to do all of this with the barrier task but only once and not for every worker.
I would assume all of this only matters for very large clusters otherwise I don't think this is a game changer.
Bottom line, I'm fine with the barrier task and think we should keep it since all of the above would still be unnecessary even if it is not sizeable (imo, we don't need to argue how sizeable it actually is). If at some point it causes problems we should still be open to throw it away again.

Reply:
> We could just as well trigger it on `remove_worker` and there is no problem.

We could. And I think for shuffling, we actually need to (even without the setup task). As a generic API, it feels a little weird to restart the tasks even when none of the tasks have been affected though. It would be so specially tuned for shuffling at that point, maybe it just shouldn't be a generic API.

> This is not a huge burden for the scheduler.

Good to know. Agreed that it's probably _more_ of a burden than the barrier, but if the difference isn't huge, maybe it doesn't matter. Still, the barrier task doesn't really cause problems currently, so I think I'll probably stick with it for now.